Assets/MixedRealityToolkit.SDK/Experimental/Features/Dwell/DwellHandler.cs

```diff
@@ -0,0 +1,243 @@
+                    FillTimer -= Time.deltaTime;
+                    FillTimer = FillTimer < 0 ? 0 : FillTimer;
+                    break;
+                case DwellStateType.DwellCompleted:
+                    Debug.Assert(FillTimer < float.Epsilon, $"Fill timer was not 0. It was {FillTimer}.");
+                    break;
+                case DwellStateType.DwellCanceled:
+                    if ((DateTime.UtcNow - focusExitTime).TotalSeconds > dwellProfile.TimeToAllowDwellResume)
+                    {
+                        FillTimer = 0;
+                        CurrentDwellState = DwellStateType.None;
+                    }
+                    break;
+                default:
+                    FillTimer = 0;
+                    break;
+            }
+        }
+
+        public void OnFocusEnter(FocusEventData eventData)
+        {
+            if (eventData.NewFocusedObject == gameObject
+                && eventData.Pointer.InputSourceParent.SourceType == dwellProfile.DwellPointerType)
+            {
+                HasFocus = true;
+
+                // check intent to resume
+                if (CurrentDwellState == DwellStateType.DwellCanceled
+                    && pointer.InputSourceParent.SourceId == eventData.Pointer.InputSourceParent.SourceId //make sure the returning pointer id is the same
+                    && (DateTime.UtcNow - focusExitTime).TotalSeconds <= dwellProfile.TimeToAllowDwellResume)
+                {
+                    focusEnterTime = focusEnterTime.AddSeconds(dwellProfile.TimeToAllowDwellResume);
+                    CurrentDwellState = DwellStateType.DwellStarted;
+                    DwellStarted.Invoke(pointer);
+                }
+                // dwell state machine re-starts
+                if (CurrentDwellState <= DwellStateType.DwellIntended)
+                {
+                    focusEnterTime = DateTime.UtcNow;
+                    CurrentDwellState = DwellStateType.FocusGained;
+                    pointer = eventData.Pointer;
+                    FillTimer = 0;
+                }
+            }
+        }
+
+        public void OnFocusExit(FocusEventData eventData)
+        {
+            if (eventData.OldFocusedObject == gameObject
+                && eventData.Pointer.InputSourceParent.SourceType == dwellProfile.DwellPointerType
+                && pointer.InputSourceParent.SourceId == eventData.Pointer.InputSourceParent.SourceId)
+            {
+                HasFocus = false;
+
+                if (CurrentDwellState == DwellStateType.DwellStarted)
+                {
+                    DwellCanceled.Invoke(eventData.Pointer);
+                    CurrentDwellState = DwellStateType.DwellCanceled;
+                    focusExitTime = DateTime.UtcNow;
+                }
+                else
+                {
+                    CurrentDwellState = DwellStateType.None;
+                    focusExitTime = DateTime.MinValue;
```

Comment:
>MinValue [](start = 45, length = 8)

Out of curiousity, why does this use MinValue (even though the default "no value" is MaxValue when it gets initialized way above?)

Reply:
my oversight. :(